### ISSUE TITLE
Rearrange order of Snippets in ModelViewSet index view

### ISSUE THREAD

**Issue by Timusan** [wagtail/wagtail#10816]
### Is your proposal related to a problem?

With the deprecation of ModelAdmin in favor of Snippets we lose the functionality delivered by [Wagtail Orderable](https://github.com/elton2048/wagtail-orderable): Ordering Snippets in the Index view by the use of drag & drop.

### Describe the solution you'd like

As is already the case for a listing of Page models (using the "Sort menu order" feature), enable Snippets to be ordered by drag & drop. It seems Wagtail admin's core already has the functionality build in. Enabling custom sort order this way would, as with Pages, negate the sorting columns. Clicking on a header column for sort disables the drag & drop again. 

The above module, that does this for ModelAdmin, adds in a `sort_order` field to do its bidding, yet extending the functionality that the Pages Index already uses would fit more as no third-party package is needed.

**Comment by abhinav700** [comment]
@Timusan I would like to work on this issue. Can you please assign it to me.

**Comment by Timusan** [comment]
@abhinav700 Awesome that you want to work on this! However, I am not the maintainer of the Wagtail repo so I am unable to assign tickets to someone I'm afraid.

**Comment by lb-** [comment]
@abhinav700 - I'm not sure that this functionality is suited to Wagtail core. It would need some discussion on how you intend to implement this before any kind of PR or Dev starts. 

I think the best solution would be for the existing package, or a new package, be managed by the community for this kind of thing.

There is some functionality for ordering in the admin (pages), maybe it's worth exploring making that JS code reusable with migrating it to Stimulus. This would pave the way for part of this functionality to be usable in Snippets. 

If you're keen on contributing that JavaScript change, I can raise a new issue that specifies what's needed.

**Comment by abhinav700** [comment]
@lb- Yeah, I would like to contribute that javascript change. Can you please raise that issue?

**Comment by lb-** [comment]
@abhinav700  - https://github.com/wagtail/wagtail/issues/10909 created.

**Comment by gasman** [comment]
Now that #10909 is in, we're in a good position to consider what the server-side code for this would look like. Proposed outline:

* TBD: decide whether this functionality will be specific to snippets, or available to all ModelViewSet-registered models. If we're following the same UI as pages, then the reordering handles will replace the bulk actions checkboxes, and currently those are snippet-specific. However, the below assumes we're applying this to ModelViewSet as a whole.
* `ModelViewSet` accepts an optional `sort_order_field` parameter; if the model defines a `sort_order_field` attribute (the `wagtail.models.Orderable` mixin does this, although as far as I can see it's not currently used for anything), `ModelViewSet` will automatically pick that up.
* If `sort_order_field` is specified, `ModelViewSet` will pass this on to its Index and Create views, and register an additional 'reorder' view to handle the POST requests made by the reordering interface on the index.
* The reordering operation will be available to users with 'publish' permission for models that extend `DraftStateMixin`, and 'change' permission for models that do not.
* `wagtail.admin.views.generic.models.CreateView` will be updated so that whenever `sort_order_field` is defined, newly-created instances will have that field populated with the next available sort order index.
* `wagtail.admin.views.generic.models.IndexView` will be updated so that whenever `sort_order_field` is defined, an item is added to the top action menu allowing users with the appropriate permission to set ordering. When this is activated, the bulk actions checkboxes will be replaced with the `w-orderable` drag-to-reorder UI.

**Comment by rjbullock** [comment]
Has there been any progress on this project? We used Wagtail Orderable quite a bit but have lost the functionality by migrating to SnippetViewSets.

**Comment by thibaudcolas** [comment]
Merged in #13323, for release in Wagtail 7.2! Release candidate due later this week for anyone who wants to test this.

Thank you @joeyjurjens for getting this started in #12857! And @laymonage taking over in #13323. Possible follow-up tasks:

- #13483 
- #13484 
- #13485
